### PR TITLE
feat: NetworkManager hibernate command, NetworkHandle hibernate funct…

### PR DESCRIPTION
**Pull request by loocapro**
Related to #1208

Hi @mattsse,

I hope i am in the right direction.

Leaving stages order and Offline stages details for reference below.

Stages execution order:

1. HeaderStage
2. TotalDifficultyStage
3. BodyStage
4. SenderRecoveryStage
5. ExecutionStage
6. MerkleStage (unwind)
7. AccountHashingStage
8. StorageHashingStage
9. MerkleStage (execute)
10. TransactionLookupStage
11. IndexStorageHistoryStage
12. IndexAccountHistoryStage
13. FinishStage

**Offline Stages**

**ExecutionStages**
 SenderRecoveryStage, ExecutionStage
 
**HashingStages**
MerkleStage, AccountHashingStage, StorageHashingStage, MerkleStage

**HistoryIndexingStages**
TransactionLookupStage, IndexStorageHistoryStage, IndexAccountHistoryStage

### CODE REVIEW

**Review by mattsse — CHANGES_REQUESTED** [review]
this looks good, though here's no way to rewake it yet, so we need that as well, 

I think we want an enum for activate,hibernate

and I don't want to integrate this in the stages at all because it's just one more thing to keep track of and not really useful, because tx gossip during sync is already disabled.

**Comment by loocapro** [comment]
> during initial sync, once bodies have been downloaded we no longer require the network (peers) since all operations are IO+Disk bound.

> and I don't want to integrate this in the stages at all because it's just one more thing to keep track of and not really useful, because tx gossip during sync is already disabled.

@mattsse what should be the recommended way to hibernate the network once the BodyStage is done but without integrating it inside the stages?

**Comment by mattsse** [comment]
I don't really want to touch that in the stages, but I can see how this feature could be useful for folks that use the network crate as a standalone component

**Comment by loocapro** [comment]
> I don't really want to touch that in the stages, but I can see how this feature could be useful for folks that use the network crate as a standalone component

Oh ok! I was looking at pipeline events actually. Do you think it would make sense to plug the Hibernate logic there? Or do you already have some logic that needs to be executed right after/before a specific stage?

**Review by mattsse — CHANGES_REQUESTED** [review]
okay after having a closer look at this, I'd like to change a few small things:

move:

https://github.com/paradigmxyz/reth/blob/matt%2Fnode-builder/crates/net/network/src/swarm.rs#L60-L61

to `PeersManager` and use this to track the state so we don't need the extra bool there

**Comment by loocapro** [comment]
> okay after having a closer look at this, I'd like to change a few small things:
> 
> move:
> 
> https://github.com/paradigmxyz/reth/blob/matt%2Fnode-builder/crates/net/network/src/swarm.rs#L60-L61
> 
> to `PeersManager` and use this to track the state so we don't need the extra bool there

@mattsse, thanks for this, i have followed your comments!

How about handling NetworkConnectionState during pipeline events to toggle the filling of outbound connection slots? Too critical?

File: crates/net/network/src/swarm.rs [hunk 1]
```diff
@@ -439,9 +434,21 @@ pub(crate) enum SwarmEvent {
 /// When in hibernation mode, the node will not initiate new outbound connections. This is
 /// beneficial for sync stages that do not require a network connection.
 #[derive(Debug, Default)]
-pub(crate) enum NetworkConnectionState {
+pub enum NetworkConnectionState {
     #[default]
     Active,
     ShuttingDown,
     Hibernate,
 }
```
Comment (mattsse) [inline crates/net/network/src/swarm.rs]: now that these are public, we should add some docs here

File: crates/net/network/src/network.rs [hunk 2]
```diff
@@ -432,6 +439,8 @@ pub(crate) enum NetworkHandleMessage {
     GetReputationById(PeerId, oneshot::Sender<Option<Reputation>>),
     /// Initiates a graceful shutdown of the network via a oneshot sender.
     Shutdown(oneshot::Sender<()>),
+    /// Toggles the network state between hibernation and active.
+    ToggleNetworkConnection(NetworkConnectionState),
```
Comment (mattsse) [inline crates/net/network/src/network.rs:443]:
```suggestion
    SetNetworkState(NetworkConnectionState),
```

File: crates/net/network/src/network.rs [hunk 3]
```diff
@@ -152,6 +153,12 @@ impl NetworkHandle {
         rx.await
     }
 
+    /// Toggles network connection state between [`NetworkConnectionState::Hibernate`] and
+    /// [`NetworkConnectionState::Active`].
+    pub async fn toggle_network_conn(&self, network_conn: NetworkConnectionState) {
+        self.send_message(ToggleNetworkConnection(network_conn));
+    }
```
Comment (mattsse) [inline crates/net/network/src/network.rs]: here we want to flatten the enum into individual commands, be we can keep this as well, the other calls simply call this then

File: crates/net/network/src/manager.rs [hunk 4]
```diff
@@ -551,6 +551,23 @@ where
...
+                // connection.
+                match net_state {
+                    NetworkConnectionState::Hibernate => {
+                        self.swarm.on_hibernate_requested();
+                    }
+                    NetworkConnectionState::Active => {
+                        self.swarm.on_network_active();
+                    }
+                    // It can only trigger a state change between Active and Hibernate.
+                    _ => {}
```
Comment (mattsse) [inline crates/net/network/src/manager.rs:567]: we should still set the state,

Comment (mattsse) [inline crates/net/network/src/manager.rs:567]: this should happen in the peers manager, and simply do 

self.state = state;

**Review by mattsse — CHANGES_REQUESTED** [review]
almost there

**Review by mattsse — APPROVED** [review]
thanks, lgtm

I only moved the hibernate check out of the loop and added some additional docs

**Pull request merged**